lib/tests/streamlit/state/session_state_test.py

```diff
@@ -185,3 +186,171 @@ def test_setting_before_widget_creation(self):
 
         c = st.checkbox("checkbox", key="c")
         assert c == True
+
+
+class SessionStateTests(unittest.TestCase):
+    def setUp(self):
+        old_state = {"foo": "bar", "baz": "qux", "corge": "grault"}
+        new_session_state = {"foo": "bar2"}
+        new_widget_state = {"baz": "qux2", f"{GENERATED_WIDGET_KEY_PREFIX}-foo": "bar"}
+        self.session_state = SessionState(
+            old_state, new_session_state, new_widget_state
+        )
+
+    def test_compact_state(self):
+        self.session_state.compact_state()
+        assert self.session_state._old_state == {
+            "foo": "bar2",
+            "baz": "qux2",
+            "corge": "grault",
+            f"{GENERATED_WIDGET_KEY_PREFIX}-foo": "bar",
+        }
+        assert self.session_state._new_session_state == {}
+        assert self.session_state._new_widget_state == {}
+
+    def test_clear_state(self):
+        self.session_state.clear_state()
+        assert self.session_state._merged_state == {}
+
+    def test_merged_state(self):
+        assert self.session_state._merged_state == {
+            "foo": "bar2",
+            "baz": "qux2",
+            "corge": "grault",
+            f"{GENERATED_WIDGET_KEY_PREFIX}-foo": "bar",
+        }
+
+    def test_filtered_state(self):
+        assert self.session_state.filtered_state == {
+            "foo": "bar2",
+            "baz": "qux2",
+            "corge": "grault",
+        }
+
+    def is_new_state_value(self):
+        assert self.session_state.is_new_state_value("foo")
+        assert not self.session_state.is_new_state_value("corge")
+
+    def test_getitem(self):
+        assert self.session_state["foo"] == "bar2"
+
+    def test_getitem_error(self):
+        with pytest.raises(KeyError):
+            self.session_state["nonexistent"]
+
+    def test_setitem(self):
+        assert not self.session_state.is_new_state_value("corge")
+        self.session_state["corge"] = "grault2"
+        assert self.session_state["corge"] == "grault2"
+        assert self.session_state.is_new_state_value("corge")
+
+    def test_setitem_disallows_setting_created_widget(self):
+        mock_ctx = MagicMock()
+        mock_ctx.widget_ids_this_run = _StringSet()
+        mock_ctx.widget_ids_this_run.add("widget_id")
+
+        with patch("streamlit.report_thread.get_report_ctx", return_value=mock_ctx):
+            with pytest.raises(StreamlitAPIException) as e:
+                self.session_state["widget_id"] = "blah"
+            assert "is disallowed." in str(e.value)
+
+    def test_delitem(self):
```

Comment:
Rather than testing that each state dict no longer has it, I would prefer we test just that it no longer appears in `self.session_state`. That way the test doesn't care about the implementation, just the behavior it promises as a MutableMapping.

Reply:
:+1: done